test/AutoDiff/Sema/derivative_attr_type_checking.swift

```diff
@@ -539,24 +549,14 @@ extension HasStoredProperty {
 // Test cross-file derivative registration. Currently unsupported.
 // TODO(TF-1021): Lift this restriction.
 
-extension AdditiveArithmetic where Self: Differentiable {
+extension FloatingPoint where Self: Differentiable {
   // expected-error @+1 {{derivative not in the same file as the original function}}
```

Comment:
I think a standalone test for "`@derivative` of protocol requirement produces error" is missing.

The test here for `@derivative(FloatingPoint.rounded)` fits the bill, except that cross-file derivative registration is currently disabled so the error message is misleading (`derivative not in the same file as the original function`). The test is also not standalone (depends on stdlib protocol/method).

Here's a standalone test:

```swift
// Test derivative registration for protocol requirements.
// TODO(TF-982): Enable this properly.
protocol ProtocolRequirementDerivative {
  func foo(_ x: Float) -> Float
}
extension ProtocolRequirementDerivative {
  // NOTE: the error is misleading because `findAbstractFunctionDecl` in
  // TypeCheckAttr.cpp is not setup to show customized error messages for
  // invalid original function candidates.
  // expected-error @+1 {{could not find function 'foo' with expected type '<Self where Self : ProtocolRequirementDerivative> (Self) -> (Float) -> Float'}}
  @derivative(of: foo)
  func vjpFoo(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
    fatalError()
  }
}
```

I'll help commit this test now.

Reply:
Done in https://github.com/apple/swift/pull/29031/commits/29465f82add36c3a964e2f8bd1557c7d35934f81.